rfcs/2021-05-05-7351-framing-and-codecs.md

```diff
@@ -0,0 +1,183 @@
+# RFC 7351 - 2021-05-05 - Framing and Codecs
+
+This RFC discusses a standardized way of specifying framing and codecs across
+sources and sinks. _Framing_ is concerned with turning sequences of bytes into
+byte frames (which indicate the boundaries of a complete message). Codecs
+consist of _decoders_, which deserialize a byte frame into structured data, and
+_encoders_, which serialize structured data into a byte frame.
+
+Conceptually, we want reusable pieces of logic which would allow us to collapse
+a source + (decoder) transform into a source with decoder config, and to
+collapse an (encoder) transform + sink into a sink with encoder config, in a way
+that is transparent to the user. These transforms merely convert between
+encoding formats, e.g. from `bytes` to `json` and vice versa.
+
+## Scope
+
+The scope of this RFC concern at which level framing and decoding operates, how
+these framers and codecs can be configured, and how they can be shared in a
+uniform way. It does not cover any specific implementation for framing or a
+codec. Also see [future work](#future-work) for extended goals that are not
+covered by this RFC.
+
+## Motivation
+
+Currently, we have no explicit abstraction that is responsible for handling
+framing and codecs, such that each source and sink may implement them in their
+own way. This can introduce inconsistencies and additional maintenance burden.
+Components have their own defaults, making behavior unpredictable and surprising
+users, e.g. as documented in
+[#3453](https://github.com/timberio/vector/issues/3453).
+
+## Internal Proposal
+
+Since decoding and encoding should apply universally to all sources and sinks,
+it should be configurable by a shared field. This is accomplished by the
+`SourceOuter` and `SinkOuter` wrappers:
```

Comment:
This also enables topology and reload level implementation/changes to be quite minimal since then we can treat them as a part of source/sink for which we have a precedent with `buffers`.